pinot-core/src/main/java/org/apache/pinot/core/segment/processing/framework/SegmentMapper.java

```diff
@@ -0,0 +1,148 @@
+import org.apache.pinot.core.segment.processing.partitioner.PartitionerFactory;
+import org.apache.pinot.core.segment.processing.transformer.RecordTransformer;
+import org.apache.pinot.core.segment.processing.transformer.RecordTransformerFactory;
+import org.apache.pinot.core.segment.processing.utils.SegmentProcessorUtils;
+import org.apache.pinot.spi.data.readers.GenericRow;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Mapper phase of the SegmentProcessorFramework.
+ * Reads the input segment and creates partitioned avro data files
+ * Performs:
+ * - record transformations
+ * - partitioning
+ * - partition filtering
+ */
+public class SegmentMapper {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentMapper.class);
+  private final File _inputSegment;
+  private final File _mapperOutputDir;
+
+  private final String _mapperId;
+  private final Schema _avroSchema;
+  private final RecordTransformer _recordTransformer;
+  private final Partitioner _partitioner;
+  private final PartitionFilter _partitionFilter;
+  private final Map<String, DataFileWriter<GenericData.Record>> _partitionToDataFileWriterMap = new HashMap<>();
+
+  public SegmentMapper(String mapperId, File inputSegment, SegmentMapperConfig mapperConfig, File mapperOutputDir) {
+    _inputSegment = inputSegment;
+    _mapperOutputDir = mapperOutputDir;
+
+    _mapperId = mapperId;
+    _avroSchema = SegmentProcessorUtils.convertPinotSchemaToAvroSchema(mapperConfig.getPinotSchema());
+    _recordTransformer = RecordTransformerFactory.getRecordTransformer(mapperConfig.getRecordTransformerConfig());
+    _partitioner = PartitionerFactory.getPartitioner(mapperConfig.getPartitioningConfig());
+    _partitionFilter = PartitionerFactory.getPartitionFilter(mapperConfig.getPartitioningConfig());
+    LOGGER.info(
+        "Initialized mapper with id: {}, input segment: {}, output dir: {}, recordTransformer: {}, partitioner: {}, partitionFilter: {}",
+        _mapperId, _inputSegment, _mapperOutputDir, _recordTransformer.getClass(), _partitioner.getClass(),
+        _partitionFilter.getClass());
+  }
+
+  /**
+   * Reads the input segment and generates partitioned avro data files into the mapper output directory
+   * Records for each partition are put into a directory of its own withing the mapper output directory, identified by the partition name
+   */
+  public void map()
+      throws Exception {
+
+    PinotSegmentRecordReader segmentRecordReader = new PinotSegmentRecordReader(_inputSegment);
+    GenericRow reusableRow = new GenericRow();
+    GenericData.Record reusableRecord = new GenericData.Record(_avroSchema);
+
+    Set<String> selectedPartitions = new HashSet<>();
+    Set<String> rejectedPartitions = new HashSet<>();
+
+    while (segmentRecordReader.hasNext()) {
+      reusableRow = segmentRecordReader.next(reusableRow);
+
+      // Record transformation
+      reusableRow = _recordTransformer.transformRecord(reusableRow);
```

Comment:
How do we handle `record filtering` that happens in the `transformRecord`?

Or, are we model filtering as a separate step? We should have some convention on record filtering.

Reply:
Currently no record filtering happens in transformRecord. If we want to do record filtering in Mapper, I think we can introduce a new phase, then it would become  transform -> filter -> partition -> partitionFilter -> flush

My goal in this phase was to mainly include everything that the current code (of segment merge/rollup) does, and also include everything that the realtime to offline project needs. Filtering wasn't required in either, so skipped it

Reply:
I have removed PartitionFilter step and introduced record filtering.